shutter/shutter.md

```diff
@@ -0,0 +1,284 @@
+
+### Interfaces with
+
+- Keypers: To receive keys from
+- Validators: To send keys to
+
+### Interface
+
+The relay provides an HTTP GET endpoint at path `/v1/decryption_keys` with the following required query parameters:
+
+- `instance_id`: A decimal encoded instance ID
+
+If the relay is connected to a gossip network with different instance ID, it responds with status code `404`. Otherwise, an SSE stream is opened that yields items of the following format:
+
+```jsx
+event: decryption_key;
+data: data;
+```
+
+`data` is the `0x`-prefixed, hex encoded decryption key protocol buffer defined under “Keypers”.
+
+As the relay is meant to be an only provisional solution it does not perform any authentication.
+
+## Encodings
+
+This section describes how various data types used above are canonically encoded.
+
+### P2P Messages
+
+Gossiped p2p messages are encoded as one of the following protocol buffers:
+
+```jsx
+message DecryptionKeyShare {
+    uint64 instanceID = 1;
+    uint64 eon = 2;
+    bytes epochID = 3;
+    uint64 keyperIndex = 4;
+    bytes share = 5;
+}
+
+message DecryptionKey {
+    uint64 instanceID = 1;
+    uint64 eon = 2;
+    bytes epochID = 3;
+    bytes key = 4;
+}
+```
+
+- `instanceID`: The ID of the instance for which the message is relevant.
+- `eon`: An eon index.
+- `epochID`: A big-endian encoded slot number. Note that this field is unrelated to the epochs of the beacon chain.
+- `keyperIndex`: A zero-based number indexing a keyper set.
+- `key`: A decryption key encoded as described below.
+- `share`: A decryption key share encoded as described below.
+
+Messages are wrapped in the following envelope:
+
+```jsx
+message Envelope {
+    uint32 version = 1 ;
+    google.protobuf.Any message = 2;
+}
+```
+
+- `version`: `0`, to be incremented with future incompatible protocol changes.
+- `message`: Any of the above defined message types.
+
+### Cryptographic Objects
+
+The cryptographic objects described below are composed of elements of the bilinear group BN256. Consequently, their encoding functions are composed of the encoding functions of these primitives.
+
+The relevant primitives are points on the curves G1 and G2. They are encoded according to [EIP-197](https://eips.ethereum.org/EIPS/eip-197#encoding).
+
+#### Decryption Keys
+
+Decryption keys are points on G1 and are encoded accordingly.
+
+#### Decryption Key Shares
+
+Decryption key shares are points on G1 and are encoded accordingly.
+
+#### Encrypted Messages
+
+Encrypted messages are tuples `(C1, C2, C3)` where
+
+- `C1` is a point on G2,
+- `C2` is a 32 bytes block, and
+- `C3` is a sequence of one or more 32 bytes blocks.
+
+Its encoding is the concatenation of its part, i.e. `encode_g2(C1) | C2 | C3[0] | ... | C3[n]`.
```

Comment:
Where is `encode_g2` defined? I don't find it in https://eips.ethereum.org/EIPS/eip-197#encoding either even then please redefine here